### PR TITLE
feat(compaction): support recreate block stream for fast compact iter

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Before I finish rebuilding the stream at the object store level, I want to make sure that fast compaction is robust by using a workaround.

So, this PR has a few changes.

1. support recreate stream for fast compact iter
2. fix remote_io metrics
3. add fail point test

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/hummock/compactor/fast_compactor_runner.rs [hunk 1]
```diff
@@ -153,9 +235,14 @@ impl BlockStreamIterator {
         }
     }
 
-    fn is_valid(&self) -> bool {
+    pub(crate) fn is_valid(&self) -> bool {
         self.iter.is_some() || self.next_block_index < self.sstable.meta.block_metas.len()
     }
+
+    #[cfg(feature = "failpoints")]
+    pub(crate) fn mut_iter(&mut self) -> &mut BlockIterator {
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/fast_compactor_runner.rs:243]: nits: by rust convention, we usually call this iter_mut

File: src/storage/src/hummock/compactor/fast_compactor_runner.rs [hunk 2]
```diff
@@ -76,34 +84,108 @@ impl BlockStreamIterator {
...
+                Some(block_stream) => block_stream.next_block_impl().await,
+                None => {
+                    self.create_stream().await?;
+                    continue;
+                }
+            };
+            match ret {
+                Ok(Some((data, _))) => {
+                    let meta = self.sstable.meta.block_metas[self.next_block_index].clone();
+                    let filter_block = if !cfg!(feature = "failpoints") {
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/fast_compactor_runner.rs:142]: why do we need a special check here? Is it too complicated to fix the test?

**Review by hzxa21 — APPROVED** [review]
Rest LGTM! Thanks for the fix.

File: src/storage/src/hummock/compactor/fast_compactor_runner.rs (hunk 2 above)
Comment (Li0k) [reply src/storage/src/hummock/compactor/fast_compactor_runner.rs:142]: just a workaround for failpoints test. The `test_utils` generate fake sst without bloom filter, I will try to remove it tmr.

**Comment by Li0k** [comment]
https://github.com/risingwavelabs/risingwave/issues/15973

**Pull request merged**